// @sudoku/stores/stateManager.js
import { SUDOKU_SIZE } from '@sudoku/constants';
import { encodeOriginSudoku } from '@sudoku/sencode/index'

function createStateManager() {
    const state_dict = new Map();   // key:index value:grid_state, grid_state是encode的字符串
    const index_dict = new Map();   // key:grid value:index

    return {
        add_state(state) {
            let newState = encodeOriginSudoku(state);

            if(!index_dict.has(newState)) {    // 没有则插入
                let index = state_dict.size;
                state_dict.set(index, newState);
                index_dict.set(newState, index);
            }
            else {
                console.log('the state exist!');
            }
        }, 
        get_index(state) {
            return index_dict.get(encodeOriginSudoku(state));
        },
        get_state(index) {
            return state_dict.get(index);
        },
        get_state_dict() {
            return state_dict;
        },
        get_index_dict() {
            return index_dict;
        },
        clear() {
            state_dict.clear();
            index_dict.clear();
        }
    }
}

export const stateManager = createStateManager();
